include/util/indexed_data.hpp

```diff
@@ -0,0 +1,366 @@
+                               Offset data_offset,
+                               OffsetIterator first,
+                               OffsetIterator last) const
+    {
+        BOOST_ASSERT(data_offset <= std::numeric_limits<decltype(BlockReference::offset)>::max());
+
+        Offset prefix_length = 0;
+        BlockReference refernce{static_cast<decltype(BlockReference::offset)>(data_offset), 0};
+        for (; first != last; --last)
+        {
+            const std::uint32_t data_length = *last - *std::prev(last);
+            if (data_length >= 0x1000000)
+                throw util::exception(boost::format("too large data length %1%") % data_length);
+
+            const std::uint32_t byte_length = log256(data_length);
+            refernce.descriptor = (refernce.descriptor << 2) | byte_length;
+            prefix_length += byte_length;
+        }
+
+        out.write((const char *)&refernce, sizeof(refernce));
+
+        return prefix_length;
+    }
+
+    /// Write a block prefix that is an array of variable encoded data lengths:
+    ///   0 is omitted;
+    ///   1..255 is 1 byte;
+    ///   256..65535 is 2 bytes;
+    ///   65536..16777215 is 3 bytes.
+    /// [first..last] is an inclusive range of block data.
+    /// The length of the last item in the block is not stored.
+    template <typename OffsetIterator>
+    void WriteBlockPrefix(std::ostream &out, OffsetIterator first, OffsetIterator last) const
+    {
+        for (OffsetIterator curr = first, next = std::next(first); curr != last; ++curr, ++next)
+        {
+            const std::uint32_t data_length = *next - *curr;
+            const std::uint32_t byte_length = log256(data_length);
+            if (byte_length == 0)
+                continue;
+
+            out.write((const char *)&data_length, byte_length);
+        }
+    }
+
+    /// Advances the range to an item stored in the referenced block.
+    /// Input [first..last) is a range of the complete block data with prefix.
+    /// Output [first..last) is a range of the referenced data at local_index.
+    template <typename DataIterator>
+    void ReadRefrencedBlock(const BlockReference &reference,
+                            std::uint32_t local_index,
+                            DataIterator &first,
+                            DataIterator &last) const
+    {
+        std::uint32_t descriptor = reference.descriptor;
+        DataIterator var_lengths = first;          // iterator to the variable lengths part
+        std::advance(first, sum2bits(descriptor)); // advance first to the block data part
+        for (std::uint32_t i = 0; i < local_index; ++i, descriptor >>= 2)
+        {
+            var_advance(first, var_lengths, descriptor & 0x3);
+        }
+
+        if (local_index < BLOCK_SIZE)
+        {
+            last = first;
+            var_advance(last, var_lengths, descriptor & 0x3);
+        }
+    }
+};
+
+template <int N, typename T = std::string> struct FixedGroupBlock
```

Comment:
It can be useful in some cases. Originally i thought to use `FixedGroupBlock` for `util::RangeTable<> bearing_class_range_table(bearing_counts)`, but did not manage to finish it.